// mgr-swagger-express/dist/utils.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractContextFromToken = function (token) {
    var splitted = token.split(';');
    return {
        author: splitted[0],
        organization: splitted[1],
        roles: splitted[2].split(','),
    };
};
exports.transformURLParameters2Swagger = function (parameters) { return parameters
    ? (parameters.map(function (param) { return (__assign(__assign({}, param), { in: 'path', type: param.type || 'string' })); }))
    : []; };
exports.transformBody2Swagger = function (body) { return body
    ? [__assign(__assign({}, body), { in: 'body' })]
    : []; };
exports.transformQuery2Swagger = function (query) { return query
    ? query.items.map(function (conf) { return ({
        in: 'query',
        name: conf.name,
        description: query.description,
        required: conf.required,
    }); })
    : []; };
